rs/ledger_suite/icrc1/ledger/src/main.rs

```diff
@@ -225,11 +233,68 @@ fn post_upgrade(args: Option<LedgerArgument>) {
 
     PRE_UPGRADE_INSTRUCTIONS_CONSUMED.with(|n| *n.borrow_mut() = pre_upgrade_instructions_consumed);
 
+    Access::with_ledger_mut(|ledger| {
+        ledger.state = LedgerState::Migrating(LedgerField::Allowances);
+        ledger.clear_arrivals();
+    });
+    migrate_next_part(
+        MAX_INSTRUCTIONS_PER_UPGRADE.saturating_sub(pre_upgrade_instructions_consumed),
+    );
+
     let end = ic_cdk::api::instruction_counter();
     let instructions_consumed = end - start;
     POST_UPGRADE_INSTRUCTIONS_CONSUMED.with(|n| *n.borrow_mut() = instructions_consumed);
 }
 
+fn migrate_next_part(instruction_limit: u64) {
+    STABLE_UPGRADE_MIGRATION_STEPS.with(|n| *n.borrow_mut() += 1);
+    let mut migrated_allowances = 0;
+    let mut migrated_expirations = 0;
+
+    ic_cdk::println!("Migration started.");
+    log!(&LOG, "Migration started.");
+
+    Access::with_ledger_mut(|ledger| {
+        while instruction_counter() < instruction_limit {
+            let field = match ledger.state.clone() {
+                LedgerState::Migrating(ledger_field) => ledger_field,
+                LedgerState::Ready => break,
+            };
+            match field {
+                LedgerField::Allowances => {
+                    if ledger.migrate_one_allowance() {
+                        migrated_allowances += 1;
+                    } else {
+                        ledger.state = LedgerState::Migrating(LedgerField::AllowancesExpirations);
+                    }
+                }
+                LedgerField::AllowancesExpirations => {
+                    if ledger.migrate_one_expiration() {
+                        migrated_expirations += 1;
+                    } else {
+                        ledger.state = LedgerState::Ready;
+                    }
+                }
+            }
+        }
+        let msg = format!("Number of elements migrated: allowances:{migrated_allowances} expirations:{migrated_expirations}. Instructions used {}.",
```

Comment:
The reported instructions used cover all instructions in the current message, not just instructions spent in the migration itself.

Reply:
Changed to "Total instructions used in message {}."

Reply:
Do we also want to collect instructions used in the migration itself (by taking a difference of perf counters)?

Reply:
Do you mean for the migration step itself (https://github.com/dfinity/ic/pull/1513/commits/8afc10d878552ec6930cfbd5fef9da773afd2ffe) or for the whole migration?

Reply:
Whatever you consider useful - I just wanted to bring this up.